keps/sig-network/NNNN-kubelet-probes/README.md

```diff
@@ -0,0 +1,1039 @@
+       pods on all nodes would have to accept connections from all
+       probing pods, meaning an attacker who was able to compromise
+       any probing pod would then be able to connect to any pod in the
+       cluster.
+
+     - Alternatively, the "probing pod" could be a static pod with a
+       special node-local IP, allocated independently of the normal
+       pod network. (Handwave handwave handwave...)
+
+2. Kubelet could perform probes inside pods without needing any
+   changes to the runtime, by attaching an ephemeral container to each
+   pod, containing some `kubelet-probe` binary that would run the
+   probes and report the results back to kubelet (by some means). Most
+   plausibly, this would be a single container that would persist
+   through the life of the pod, periodically running probes, rather
+   than being a new ephemeral container attached each time kubelet
+   needed to perform a probe. Kubelet could perhaps attach such a
+   container to the pod at construction time, via CRI calls, without
+   exposing the container to the kubernetes API, like it does with the
+   pause / sandbox / infrastructure container. (In fact, the probes
+   could even be done as part of the existing pause / sandbox /
+   infrastructure container.)
+
+3. If we can improve the performance of exec probes, then kubelet
+   could simply implement the network probes as exec probes, though
+   this would again require behind-the-scenes modifications to the pod
+   via CRI, to mount a volume containing the `kubelet-probe` binary so
+   that it's available to the exec probe.
+
+4. We could add a new method to CRI to allow kubelet to do the
+   equivalent of `nsenter -n` / `ip netns exec` without any of the
+   other overhead of an exec probe. It is not clear that this concept
+   would map well to Windows containers or even to
+   virtualization-based container runtimes on Linux. In the worst
+   case, it could fall back to something like (3).
+
+5. We could add a new method to CRI to allow kubelet to request that
+   the runtime perform a probe inside the container via some means
+   appropriate to the runtime. If this is not simply a wrapper around
+   exec probes, then this implies that CRI will need to understand
+   each type of probe supported by Kubernetes (i.e., it will need to
+   support every feature of `HTTPGetAction`, `TCPSocketAction` and
+   `GRPCAction`.) In the future, adding new probe types (for example,
+   a UDP probe) would require adding support to both `Pod`/kubelet and
+   to CRI.
+
+     - If CRI is going to need to understand that much of the
+       semantics of Kubernetes probes, then it might make sense to
+       just _completely_ move probes into CRI, and have CRI simply
+       report back as part of the pod status whether the pod is live,
+       ready, etc, according to its probes. This might allow for
+       better optimization of probes on the CRI side.
+
```

Comment:
Since I originally wrote this, we have started talking about [KNI](https://github.com/kubernetes/enhancements/issues/4410), and the ideas here that mention CRI could alternatively be done in terms of KNI. (Assuming that KNI is going to happen in a useful time frame, this would actually make a lot of sense, since it's another case of putting more of the details of the pod network into one place rather than having them be scattered across multiple places.)